### PR TITLE
SAK-40881 - Add OIDC flow to Sakai's LTI Advantage

### PR DESCRIPTION
**Pull request by csev**
Note that this includes a database change and will need to be merged into Sakai-19.

### CODE REVIEW

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java [hunk 1]
```diff
@@ -1530,4 +1536,45 @@ private void handleLineItemsDelete(String signed_placement, String lineItem, Htt
...
+	 * @param signed_placement
+	 * @param lineItem - Can be null
+	 * @param results
+	 * @param request
+	 * @param response
+	 */
+	private void handleOIDCAuthorization(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+		String state = (String) request.getParameter("state");
+		if ( state != null && state.trim().length() < 1 ) state = null;
```
Comment (ottenhoff) [inline basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java]: StringUtils.trimToNull does all of this hard work for you!

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java [hunk 2]
```diff
@@ -1530,4 +1536,45 @@ private void handleLineItemsDelete(String signed_placement, String lineItem, Htt
...
+		}
+
+		if ( login_hint.contains("&") || login_hint.contains("?") || ! login_hint.startsWith("/access/basiclti/site/") ) {
+			LTI13Util.return400(response, "Bad format for login_hint");
+			log.error("Bad format for login_hint");
+			return;
+		}
+
+		String redirect = login_hint;
+		redirect += "?state=" + java.net.URLEncoder.encode(state);
```
Comment (ottenhoff) [inline basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java]: This seems more clear as one line

File: basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java [hunk 3]
```diff
@@ -319,10 +345,17 @@ else if ( refId.startsWith("tool:") && refId.length() > 5 )
 						if ( tool != null ) {
 							tool.put(LTIService.LTI_SITE_ID, ref.getContext());
 						}
-						retval = SakaiBLTIUtil.postContentItemSelectionRequest(toolKey, tool, rb, contentReturn, propData);
+						String state = req.getParameter("state");
+						String oidc_endpoint = (String) tool.get(LTIService.LTI13_OIDC_ENDPOINT);
+						log.debug("State={} oidc_endpoint={}",state, oidc_endpoint);
+						if ( oidc_endpoint != null && oidc_endpoint.trim().length() > 1 &&
```
Comment (ottenhoff) [inline basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java]: if (StringUtils.isNotBlank(oidc_endpoint) && StringUtils.isNotBlank(state)) {

File: basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTIService.java [hunk 4]
```diff
@@ -311,6 +313,8 @@
     String LTI13_TOOL_PRIVATE = "lti13_tool_private";
     String LTI13_PLATFORM_PUBLIC = "lti13_platform_public";
     String LTI13_PLATFORM_PRIVATE = "lti13_platform_private";
+	String LTI13_OIDC_ENDPOINT = "lti13_oidc_endpoint";
+	String LTI13_OIDC_REDIRECT = "lti13_oidc_redirect";
```
[the remark this answers is not in the document]
Comment (csev) [reply basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTIService.java]: I will fix this.

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java (hunk 2 above)
Comment (csev) [reply basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java]: Agreed.

File: basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTIService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (csev) [reply basiclti/basiclti-api/src/java/org/sakaiproject/lti/api/LTIService.java]: I will fix.

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java (hunk 1 above)
Comment (csev) [reply basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java]: I guess.  At times, I tire of using things that are not in Java normally.  Java should fix themselves. :)

File: basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java (hunk 3 above)
Comment (csev) [reply basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java]: OK :)  Fixed two places actually.

Comment (csev) [reply basiclti/basiclti-impl/src/java/org/sakaiproject/basiclti/impl/BasicLTISecurityServiceImpl.java]: Actually 
if (StringUtils.isNotBlank(oidc_endpoint) && StringUtils.isEmpty(state)) {

**Pull request merged**